ngraph/test/backend/unary_test.hpp

```diff
@@ -0,0 +1,211 @@
+        // number of data contained in test_input & test_expected, here we
+        // sample test data repetitively to generate actual test data of required size.
+        auto static_size = shape_size(ashape.static_shape);
+        std::vector<value_type> input;
+        std::vector<value_type> expected;
+        for (size_t i = 0; i < static_size; i++)
+        {
+            input.push_back(test_input[i % test_input.size()]);
+            expected.push_back(test_expected[i % test_expected.size()]);
+        }
+
+        auto test_case = test::TestCase<TestEngine>(f);
+        test_case.template add_input<value_type>(input);
+        test_case.template add_expected_output<value_type>(ashape.static_shape, expected);
+        if (tol.using_bits)
+        {
+            if (tol.tolerance_bits >= 0)
+                test_case.run(tol.tolerance_bits);
+            else
+                test_case.run();
+        }
+        else
+        {
+            test_case.run_with_tolerance_as_fp(tol.tolerance_fp);
+        }
+    }
+
+    // overloading support passing reference function instead of expected results
+    template <typename TestEngine,
+              element::Type_t element_type,
+              typename value_type = fundamental_type_for<element_type>>
+    void test_unary(std::shared_ptr<Function> f,
+                    const std::vector<value_type>& input,
+                    value_type (*func)(value_type),
+                    ArgShape ashape = {},
+                    ArgTolerance tol = {})
+    {
+        std::vector<value_type> expected;
+        for (value_type x : input)
+            expected.push_back(func(x));
+
+        test_unary<TestEngine, element_type>(f, input, expected, ashape, tol);
+    }
+
+    // overloading support passing creator instead of ngraph function
+    // Creator is any callable of following prototype:
+    //
+    //          std::shared_ptr<Function> (const ngraph::element::Type&, const PartialShape&)
+    //
+    // unary_func() is a helper function to generate a lambda as callable on simple unary ops with
+    // optional arguments captured in the that lambda.
+    //
+    using UnaryFuncCreator =
+        std::function<std::shared_ptr<Function>(const ngraph::element::Type&, const PartialShape&)>;
+
+    template <typename OpType, typename... Args>
+    typename std::enable_if<std::is_base_of<op::Op, OpType>::value, UnaryFuncCreator>::type
+        unary_func(Args&&... args)
+    {
+        return [args...](const ngraph::element::Type& ele_type, const PartialShape& pshape) {
+            auto X = make_shared<op::Parameter>(ele_type, pshape);
+            auto Y = make_shared<OpType>(X, args...);
+            return make_shared<Function>(Y, ParameterVector{X});
+        };
+    }
+
+    template <typename TestEngine,
+              element::Type_t element_type,
+              typename Creator,
+              typename value_type = fundamental_type_for<element_type>>
+    void test_unary(Creator creator,
```

Comment:
Is this overload required? Can we call `creator` to create `Function` before `test_unary` call?

Reply:
It's a syntax sugar that allows a more compact & less redundant way of coding:

```
    test_unary<TestEngine, element::f32>(
        create_unary_function<op::Acosh>(),
        {0.f, 1.f, -1.f, 2.f, -2.f, 3.f, -3.f, 4.f, 5.f, 10.f, 100.f},
        std::acosh,
        {PartialShape::dynamic(), Shape{11});
```


Without this overload, caller needs to take extra care on passing compatible Shape & element type into two separate function calls.
```
auto func = create_unary_function<op::Acosh>()(element::f32, Shape{11});

test_unary<TestEngine, element::f32>(
      func,
      {0.f, 1.f, -1.f, 2.f, -2.f, 3.f, -3.f, 4.f, 5.f, 10.f, 100.f},
      std::acosh,
      Shape{11},
);

```

Note that create_unary_function\<op::Acosh\>() returns a callable instead of ngraph function.